notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+     "execution_count": 30,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# ingest sales from click data by \n",
+    "# - filtering source for purchase events with \"event_type\"='purchase'\n",
+    "# - adding a column for \"quantity\" with a constant value of 1\n",
+    "sql='''\n",
+    "REPLACE INTO \"sales\" OVERWRITE ALL\n",
+    "WITH \"ext\" AS (SELECT *\n",
+    "FROM TABLE(\n",
+    "  EXTERN(\n",
+    "    '{\"type\":\"http\",\"uris\":[\"http://datagen:9999/file/clicks.json\"]}',\n",
+    "    '{\"type\":\"json\"}'\n",
+    "  )\n",
+    ") EXTEND (\"time\" VARCHAR, \"user_id\" VARCHAR, \"event_type\" VARCHAR, \"client_ip\" VARCHAR, \"client_device\" VARCHAR, \"client_lang\" VARCHAR, \"client_country\" VARCHAR, \"referrer\" VARCHAR, \"keyword\" VARCHAR, \"product\" VARCHAR))\n",
+    "SELECT\n",
+    "  TIME_PARSE(\"time\") AS \"__time\",\n",
+    "  \"user_id\",\n",
+    "  \"event_type\",\n",
+    "  \"client_country\",\n",
+    "  \"product\",\n",
+    "  1 as \"quantity\"\n",
+    "FROM \"ext\"\n",
+    "WHERE \"event_type\"='purchase'  \n",
+    "PARTITIONED BY DAY\n",
+    "'''\n",
+    "display.run_task(sql)\n",
+    "print_timestamp()\n",
+    "\n",
+    "# ingest forecast data with a couple of adjustments\n",
+    "# - add column expression 'forecast' as event_type, such that we can distinguish between purchases and forecasts when using UNION ALL\n",
+    "# - change the name of the \"forecast_quantity\" column to \"quantity\" so that columns names match when using UNION ALL\n",
+    "sql='''\n",
+    "REPLACE INTO \"sales_forecast\" OVERWRITE ALL\n",
+    "WITH \"ext\" AS (SELECT *\n",
+    "FROM TABLE(\n",
+    "  EXTERN(\n",
+    "    '{\"type\":\"http\",\"uris\":[\"http://datagen:9999/file/forecast.json\"]}',\n",
+    "    '{\"type\":\"json\"}'\n",
+    "  )\n",
+    ") EXTEND (\"time\" VARCHAR, \"client_country\" VARCHAR, \"product\" VARCHAR, \"forecast_quantity\" INTEGER))\n",
+    "SELECT\n",
+    "  TIME_PARSE(\"time\") AS \"__time\",\n",
+    "  'forecast' as \"event_type\",\n",
+    "  \"client_country\",\n",
+    "  \"product\",\n",
+    "  \"forecast_quantity\" as \"quantity\"\n",
+    "FROM \"ext\"\n",
+    "PARTITIONED BY DAY\n",
+    "'''\n",
+    "display.run_task(sql)\n",
+    "print_timestamp()\n",
+    "\n",
+    "\n",
+    "# make sure both tables are available before moving on\n",
+    "sql_client.wait_until_ready('sales')\n",
+    "sql_client.wait_until_ready('sales_forecast')\n",
+    "print_timestamp()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fca62939-b82c-4a95-bab8-906c99bfb388",
+   "metadata": {},
+   "source": [
+    "### Independent Queries - App Joins Results"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "2a00800c-f0bf-4d7a-8141-fa04f0569d9c",
+   "metadata": {},
+   "source": [
+    "In order to compare forecast to sales, we will need to aggregate both at the same time granularity and with the same dimensions. \n",
```

Comment:
Use UNION ALL operations to compare forecasts and sales by X, Y, Z.

Reply:
I'm not talking about the UNION ALL approach yet. Just talking in general, about the issue that they are different datasets that need to conform to the same aggregation level before you can compare them. After that I go into the different methods one can use to achieve that.